skippkg-finder.py

```diff
@@ -0,0 +1,333 @@
+#!/usr/bin/python3
+
+import argparse
+import logging
+import sys
+
+from urllib.error import HTTPError
+
+import re
+from lxml import etree as ET
+from collections import namedtuple
+
+import osc.conf
+import osc.core
+from osc import oscerr
+from osc.util.helper import decode_it
+from osclib.core import source_file_ensure
+
+OPENSUSE = 'openSUSE:Leap:15.4'
+SLE = 'SUSE:SLE-15-SP4:GA'
+SUPPORTED_ARCHS = ['x86_64', 'i586', 'aarch64', 'ppc64le', 's390x']
+DEFAULT_REPOSITORY = 'standard'
+BINARY_REGEX = r'(?:.*::)?(?P<filename>(?P<name>.*)-(?P<version>[^-]+)-(?P<release>[^-]+)\.(?P<arch>[^-\.]+))'
+RPM_REGEX = BINARY_REGEX + r'\.rpm'
+
+META_PACKAGE = '000package-groups'
+
+makeurl = osc.core.makeurl
+http_GET = osc.core.http_GET
+http_POST = osc.core.http_POST
+http_PUT = osc.core.http_PUT
+
+
+class SkippkgFinder(object):
+    def __init__(self, opensuse_project, sle_project, print_only, verbose):
+        self.opensuse_project = opensuse_project
+        self.sle_project = sle_project
+        self.print_only = print_only
+        self.verbose = verbose
+        self.apiurl = osc.conf.config['apiurl']
+        self.debug = osc.conf.config['debug']
+
+    def is_sle_specific(self, package):
+        """
+        Return True if package is provided for SLE only or a SLE forking
+        """
+        pkg = package.lower()
+        if pkg.startswith('skelcd') or pkg.startswith('release-notes') or\
+                pkg.startswith('sle-') or pkg.startswith('sle_'):
+            return True
+        if 'sles' in pkg or 'sled' in pkg:
+            return True
+        if 'sap-' in pkg or '-sap' in pkg or pkg.startswith('sap'):
+            return True
+        if 'eula' in pkg or 'branding' in pkg:
+            return True
+        if pkg.startswith('sle15') or\
+                pkg.startswith('sles15') or\
+                pkg.startswith('suse-migration') or\
+                pkg.startswith('migrate') or\
+                pkg.startswith('kernel-livepatch') or\
+                pkg.startswith('patterns') or\
+                pkg.startswith('supportutils-plugin') or\
+                pkg.startswith('lifecycle-data-sle') or\
+                pkg.startswith('sca-patterns') or\
+                pkg.startswith('susemanager-') or\
+                pkg.startswith('kiwi-template') or\
+                pkg.startswith('desktop-data'):
+            return True
+        if pkg.endswith('bootstrap') or pkg.endswith('-caasp') or\
+                pkg.endswith('-sle'):
+            return True
+        if pkg == 'suse-build-key' or pkg == 'suse-hpc' or\
+                pkg == 'llvm' or\
+                pkg == 'txt2tags' or\
+                pkg == 'zypper-search-packages-plugin' or\
+                pkg == 'python-pymemcache' or\
+                pkg == 'python-ibus':
+            return True
```

Comment:
there is a lot of hardcoded stuff and I kind of miss rationally for bunch of them. Consider that you are promoted and someone like me need to adapt script for next version. So at least document how to obtain such list of SLE specific stuff.
Also e.g. parts with startswith can be quite error prone. Any software starting with patterns or migrate is matched.
Also that generic sap matcher can be error prone e.g. even `pinesap-grow`

Reply:
That is a pain for openSUSE release engineer to consider what SLE package should not pick to openSUSE, *none can tell*, those hardcoded stuff here can be 1) it has been agreed to fork in openSUSE which means we have different version between Leap and SLE , 2) really should not take to openSUSE, like desktop-data-SLE and such.

Yes, we do rebuild patterns in Leap, so any patterns from SLE should not take to openSUSE, migration stuff also, and of course it's true for sap stuff. We don't have `pinesap-grow` exist in SLE.